pkg/handler/collector/oci/oci.go

```
@@ -182,109 +210,286 @@ func (o *ociCollector) getTagsAndFetch(ctx context.Context, repo string, tags []
+	}
+
+	m, err := rc.ManifestGet(ctx, r)
+	if err != nil {
+		// this is a normal behavior, not an error when the digest does not have an attestation
+		// explicitly logging it as info to avoid call-stack when logging
+		logger.Infof("unable to get manifest for %v: %v", artifact, err)
+		return nil
+	}
+
+	// go through layers in reverse
+	mi, ok := m.(manifest.Imager)
+	if !ok {
+		return fmt.Errorf("reference is not a known image media type")
+	}
+	layers, err := mi.GetLayers()
+	if err != nil {
+		return err
+	}
+	for i := len(layers) - 1; i >= 0; i-- {
+		if ctx.Err() != nil {
+			return ctx.Err()
 		}
+		blob, err := rc.BlobGet(ctx, r, layers[i])
+		if err != nil {
+			return fmt.Errorf("failed pulling layer %d: %w", i, err)
+		}
+		defer blob.Close()
+		btr1, err := blob.RawBody()
+		if err != nil {
+			return fmt.Errorf("failed reading layer %d: %w", i, err)
+		}
+
+		var docType = processor.DocumentUnknown
+		var docFormat = processor.FormatUnknown
+
+		// check if artifactType is in wellKnownOCIArtifactTypes
+		if artifactType != "" {
+			if wellKnownArtifactType, ok := wellKnownOCIArtifactTypes[artifactType]; ok {
+				docType = wellKnownArtifactType.documentType
+				docFormat = wellKnownArtifactType.formatType
+			}
+		}
+
+		doc := &processor.Document{
+			Blob:   btr1,
+			Type:   docType,
+			Format: docFormat,
+			SourceInformation: processor.SourceInformation{
+				Collector: string(OCICollector),
+				Source:    artifact,
+			},
+		}
+		docChannel <- doc
+	}
+
+	return nil
+}
+
+// isDigestCollected checks if a given digest has already been collected for a given repository.
+// It returns true if the digest has been collected, false otherwise.
+func (o *ociCollector) isDigestCollected(repo string, digest string) bool {
+	collectedDigests, ok := o.checkedDigest.Load(repo)
+	if !ok {
+		o.checkedDigest.Store(repo, []string{})
+		return false
+	} else {
+		digests, ok := collectedDigests.([]string)
+		if !ok {
+			return false
```

same here, should we return an error